(SERVER-2215) Update cljs tools and use figwheel-sidecar

lazartea opened the pull request:
cljs-dashboard-widgets used older clojure, clojurescript, and cljsjs/d3 dependencies. I updated these to modern versions and added clj-parent to manage the rest of the dependencies in the project. Upgrading to d3 v4 broke some of the metrix-box animation, so I fixed that as well. Figwheel required a lein dependency to run lein figwheel, but upgrading to figwheel-sidecar allowed me to remove that dependency.

lazartea committed:
(SERVER-2215) Update cljs tools and use figwheel-sidecar

cljs-dashboard-widgets used older clojure, clojurescript, and cljsjs/d3 dependencies. I updated these to modern versions and added clj-parent to manage the rest of the dependencies in the project. Upgrading to d3 v4 broke some of the metrix-box animation, so I fixed that as well. Figwheel required a lein dependency to run lein figwheel, but upgrading to figwheel-sidecar allowed me to remove that dependency.

justinstoller commented on project.clj:
```diff
@@ -6,17 +6,20 @@
   :source-paths ["src/cljs"]
 
-  :dependencies [[org.clojure/clojure "1.7.0"]
+  :parent-project {:coords [puppetlabs/clj-parent "1.7.3"]
```
What happens when we upgrade to clj-parent 2.0.1?

lazartea replied:
The answer is apparently "everything breaks." I'll have a look into fixing dependency conflicts so we can upgrade.

justinstoller replied:
: (

justinstoller commented on dev/clj/puppetlabs/metrics/dashboard/demo/handler.clj:
```diff
@@ -58,7 +58,7 @@
           (update-timer-metrics functions-atom)
           {:status 200
-           :headers {"Content-Type" "application/json"}
+           :headers {"Content-Type" "application/json" "Access-Control-Allow-Origin" "http://localhost:3449" "Access-Control-Allow-Credentials" "true"}
```
We should split this into separate lines

lazartea commented:
@justinstoller both of those issues should be fixed now

justinstoller commented:
@lazartea can we get the README updated as well?

I think we should call out that for the figwheel server the host should be "localhost" (for the cross origin header) and the path should include "metrics.html".

There's also a "browser-repl" function mentioned in the README that appears to have always been aspirational. We should note that the functionality is now available from figwheel with `(cljs-repl)`.

lazartea commented:
@justinstoller I added those changes to the README, although I'm not sure if we want to go into detail about the figwheel-sidecar change? I updated the browser-repl mention in the README, and then I went back to ``user.clj`` and added a function so that the user wouldn't have to ``(require '[figwheel-sidecar.repl.api :as ra])`` and then call ``(ra/cljs-repl)`` to start the repl. I also updated the help command so that it would print the ``(cljs-repl)`` option as well.
